<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Objectives" data-toc-modified-id="Objectives-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Objectives</a></span></li><li><span><a href="#Example-Together" data-toc-modified-id="Example-Together-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Example Together</a></span><ul class="toc-item"><li><span><a href="#Question" data-toc-modified-id="Question-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Question</a></span></li><li><span><a href="#Considerations" data-toc-modified-id="Considerations-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Considerations</a></span></li><li><span><a href="#Loading-the-Data" data-toc-modified-id="Loading-the-Data-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Loading the Data</a></span></li><li><span><a href="#Some-Exploration-to-Better-Understand-our-Data" data-toc-modified-id="Some-Exploration-to-Better-Understand-our-Data-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Some Exploration to Better Understand our Data</a></span></li><li><span><a href="#Experimental-Setup" data-toc-modified-id="Experimental-Setup-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Experimental Setup</a></span><ul class="toc-item"><li><span><a href="#What-Test-Would-Make-Sense?" data-toc-modified-id="What-Test-Would-Make-Sense?-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>What Test Would Make Sense?</a></span></li><li><span><a href="#The-Hypotheses" data-toc-modified-id="The-Hypotheses-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>The Hypotheses</a></span></li><li><span><a href="#Setting-a-Threshold" data-toc-modified-id="Setting-a-Threshold-2.5.3"><span class="toc-item-num">2.5.3&nbsp;&nbsp;</span>Setting a Threshold</a></span></li></ul></li><li><span><a href="#$\chi^2$-Test" data-toc-modified-id="$\chi^2$-Test-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>$\chi^2$ Test</a></span><ul class="toc-item"><li><span><a href="#Setup-the-Data" data-toc-modified-id="Setup-the-Data-2.6.1"><span class="toc-item-num">2.6.1&nbsp;&nbsp;</span>Setup the Data</a></span></li><li><span><a href="#Calculation" data-toc-modified-id="Calculation-2.6.2"><span class="toc-item-num">2.6.2&nbsp;&nbsp;</span>Calculation</a></span></li></ul></li><li><span><a href="#Interpretation" data-toc-modified-id="Interpretation-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Interpretation</a></span></li></ul></li><li><span><a href="#Exercise" data-toc-modified-id="Exercise-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Exercise</a></span></li></ul></div>

In [1]:
import numpy as np
import pandas as pd

from scipy import stats
import seaborn as sns

# Objectives

- Conduct an A/B test in Python
- Interpret the results of the A/B tests for a stakeholder

# Example Together

## Question

We have data about whether customers completed sales transactions, segregated by the type of ad banners to which the customers were exposed.

The question we want to answer is whether there was any difference in sales "conversions" between desktop customers who saw the sneakers banner and desktop customers who saw the accessories banner in the month of May 2019.

## Considerations

What would we need to consider when designing our experiment?

Might include:

- Who is it that we're including in our test?
- How big of an effect would make it "worth" us seeing?
    - This can affect sample size
    - This can give context of a statistically significant result
- Other biases or "gotchas"

## Loading the Data

First let's download the data from [kaggle](https://www.kaggle.com/podsyp/how-to-do-product-analytics) via the release page of this repo: https://github.com/flatiron-school/ds-ab_testing/releases 

The code below will load it into our DataFrame:

In [2]:
# This will download the data from online so it can take
# some time (but relatively small download)

df = pd.read_csv('https://github.com/flatiron-school/ds-ab_testing/releases/download/v1.2/products_small.csv')

> Let's take a look while we're at it

In [3]:
df.head()

Unnamed: 0                          order_id  \
0     4122928  3e6c5e89fdddcaee0eed210ec2c9cadf   
1      564306  feed6203517d3abf6aab13761633174b   
2     1872289  e33d5d7941edc281646aa37763729771   
3     3616779  7c4caa8d508fa7c3bbc25f35cdd9168a   
4     5871482  12874b29bde8bbd43fb2b95735caf9e6   

                            user_id                           page_id  \
0  90d58d967eb72656e86059ec6f208092  2fdc16a09e0016555dd4da4a3fe84414   
1  08703dab1f004eabba25aacb7f0e5484  6b0a902b9b73d5a158d0119d6feb38ac   
2  bdf1d25697e21419901c94fabdafad15  9ddb7315c4357929931b48f2b3d11c62   
3  8d2f23a732c9527d95678088a3bac122  1f86cd0bea31d54a5b511b42fd19401a   
4  5a22604f8f31ae98ee1211ece3a02004  b533f7e2003418c63fd71471264c559a   

       product site_version                 time        title  target  
0  accessories      desktop  2019-03-06 08:42:47  banner_show       0  
1     sneakers       mobile  2019-04-19 18:30:45  banner_show       0  
2      company       mobile  2019-01-20 17:20:10  banner_show       0  
3     sneakers       mobile  2019-02-20 09:38:32  banner_show       0  
4     sneakers       mobile  2019-04-24 09:19:02  banner_show       0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 9 columns):
 #   Column        Non-Null Count    Dtype 
---  ------        --------------    ----- 
 0   Unnamed: 0    1000000 non-null  int64 
 1   order_id      1000000 non-null  object
 2   user_id       1000000 non-null  object
 3   page_id       1000000 non-null  object
 4   product       1000000 non-null  object
 5   site_version  1000000 non-null  object
 6   time          1000000 non-null  object
 7   title         1000000 non-null  object
 8   target        1000000 non-null  int64 
dtypes: int64(2), object(7)
memory usage: 68.7+ MB


## Some Exploration to Better Understand our Data

Lets's look at the different banner types:

In [5]:
df["product"].value_counts()

clothes             210996
company             203020
sneakers            201298
sports_nutrition    193200
accessories         191486
Name: product, dtype: int64

In [6]:
df.groupby('product')['target'].value_counts()

product           target
accessories       0         186127
                  1           5359
clothes           0         197720
                  1          13276
company           0         203020
sneakers          0         193375
                  1           7923
sports_nutrition  0         190363
                  1           2837
Name: target, dtype: int64

Let's look at the range of time-stamps on these data:

In [7]:
df["time"].min()

'2019-01-01 00:00:25'

In [8]:
df["time"].max()

'2019-05-31 23:59:21'

Let's check the counts of the different site_version values:

In [9]:
df["site_version"].value_counts()

mobile     718521
desktop    281479
Name: site_version, dtype: int64

And now check titles

In [10]:
df["title"].value_counts()

banner_show     872275
banner_click     98330
order            29395
Name: title, dtype: int64

In [11]:
len(df[df['title'] == 'order']), sum(df['target'])

(29395, 29395)

In [12]:
df.groupby('title').agg({'target': 'mean'})

target
title               
banner_click       0
banner_show        0
order              1

## Experimental Setup

We need to filter by site_version, time, and product:

In [13]:
df_AB = df.loc[(df["site_version"] == "desktop")&
               (df["time"] >= "2019-05-01")&
               ((df["product"] == "sneakers") | (df["product"] == "accessories"))]

In [14]:
df_AB.tail()

Unnamed: 0                          order_id  \
999904     5738290  ecfe9da2dbea255f05b2f526a4c663a7   
999908     7885931  8c28b45ff7fbb5d5df0b7670183689db   
999915     8099155  c613db2f19c7285f101196062c178490   
999930     6690689  6f7f32ecf6790e4eb06384b479f43162   
999999     7650805  af3a19fcceb477d2992f1771af3a9b34   

                                 user_id                           page_id  \
999904  cd7598ba5798ad6bd3f5a667ae4c217a  644ad50dd900d6735d8ceb15b71d50f8   
999908  2f28106a4db2babcfde1c4095c805632  a3d2de7675556553a5f08e4c88d2c228   
999915  97473368941b89c4716c4f389e9b8bd7  a3d2de7675556553a5f08e4c88d2c228   
999930  73f403b4843c70c41c0db294ee479582  13f395d7c86ef276f7ce52a557fd3491   
999999  515cb969971fb5e08ae0316f58e3f434  ef38d906f15db1efde867c68f1336946   

            product site_version                 time        title  target  
999904  accessories      desktop  2019-05-01 23:01:05  banner_show       0  
999908     sneakers      desktop  2019-05-26 11:19:09        order       1  
999915     sneakers      desktop  2019-05-11 09:04:46        order       1  
999930  accessories      desktop  2019-05-09 09:17:24  banner_show       0  
999999     sneakers      desktop  2019-05-20 17:58:16  banner_show       0

In [20]:
df_AB.reset_index(inplace=True, drop=True)
df_AB.head()

Unnamed: 0                          order_id  \
0     8064199  f909eb9bbf2795337dd38323b7951705   
1     7560076  97808cfd3d44fa7e8cfeb71bb5f02590   
2     5923209  71c39cc2e2ede95a7ca296d4e3e56f85   
3     8161379  e969a575bcf5b10d6676c2fab9cbf784   
4     4564531  3068b616fac2838de6a2d33a6fad60ad   

                            user_id                           page_id  \
0  5a96585c3db94eb2ae40e559cbd2e164  94b6ea2316b773afa20925d9b79d5bf8   
1  1e2db5d01051fc65fffc6c27f34400a3  48c8e3f287d814493cff887ff445f535   
2  46a6f13750d5032f366349e3e3171678  30559d0b23980e7a778c93c52ba5deca   
3  aa77bca62e9a84f5cd8d734c70afd060  70ed6e66458fda7a3164e2633d81ea07   
4  aec8ee5450418b5bc97127c2269058ca  14cbe8cf92c259f5d18310347e3cb5cb   

       product site_version                 time        title  target  
0  accessories      desktop  2019-05-19 09:26:29  banner_show       0  
1     sneakers      desktop  2019-05-03 05:34:46  banner_show       0  
2     sneakers      desktop  2019-05-19 05:04:32  banner_show       0  
3     sneakers      desktop  2019-05-29 17:51:46  banner_show       0  
4     sneakers      desktop  2019-05-04 09:33:44  banner_show       0

### What Test Would Make Sense?

Since we're comparing the frequency of conversions of customers who saw the "sneakers" banner against those who saw the "accessories" banner, we can use a $\chi^2$ test.

Note there are other hypothesis tests we can use but this should be fine since it should fit our criteria.

### The Hypotheses

$H_0$: Customers who saw the sneakers banner were no more or less likely to buy than customers who saw the accessories banner.

$H_1$: Customers who saw the sneakers banner were more or less likely to buy than customers who saw the accessories banner.

### Setting a Threshold

We'll set a false-positive rate of $\alpha = 0.05$.

## $\chi^2$ Test

### Setup the Data

We need our contingency table: the numbers of people who did or did not submit orders, both for the accessories banner and the sneakers banner. 

In [21]:
df_A = df_AB[df_AB["product"] == "accessories"]
df_B = df_AB[df_AB["product"] == "sneakers"]

In [23]:
accessories_orders = df_A["target"].sum()
sneakers_orders = df_B["target"].sum()

accessories_orders, sneakers_orders

(496, 799)

To get the numbers of people who didn't submit orders, we get the total number of people who were shown banners and then subtract the numbers of people who did make orders.

In [24]:
accessories_total = len(df_A)
sneakers_total = len(df_B)

accessories_no_orders = accessories_total - accessories_orders
sneakers_no_orders = sneakers_total - sneakers_orders

accessories_no_orders, sneakers_no_orders

(12207, 12470)

In [25]:
contingency_table = np.array([
    (accessories_orders, accessories_no_orders), 
    (sneakers_orders, sneakers_no_orders)
])

contingency_table

array([[  496, 12207],
       [  799, 12470]], dtype=int64)

### Calculation


In [26]:
stats.chi2_contingency(contingency_table)

(60.94640624666488,
 5.86498985882812e-15,
 1,
 array([[  633.38922686, 12069.61077314],
        [  661.61077314, 12607.38922686]]))

In [33]:
(496/(12207+496)) *100

3.9045894670550267

In [34]:
(799/(12470+799)) *100

6.021553998040545

This extremely low $p$-value suggests that these two groups are genuinely performing differently. In particular, the desktop customers who saw the sneakers banner in May 2019 bought at a higher rate than the desktop customers who saw the accessories banner in May 2019.

## Interpretation

In [30]:
contingency_table

array([[  496, 12207],
       [  799, 12470]], dtype=int64)

In [31]:
# Find the difference in conversion rate
accessory_CR, sneaker_CR = contingency_table[:, 0] / contingency_table.sum(axis=1)

In [32]:
print(f'Conversion Rate for accessory banner:\n\t{100*accessory_CR:.3f}%')
print(f'Conversion Rate for sneaker banner:\n\t{100*sneaker_CR:.3f}%')
print('')
print(f'Absolute difference of CR: {100*(sneaker_CR-accessory_CR):.3f}%')

Conversion Rate for accessory banner:
	3.905%
Conversion Rate for sneaker banner:
	6.022%

Absolute difference of CR: 2.117%


So we can say:
- There was a statistically significant difference at the $\alpha$-level (confidence level)
- The difference was about $2.3\%$ in favor of the sneaker banner!

# Exercise

> The company is impressed with what you found and is now wondering if there is a difference in their other banner ads!

With your group, look at the same month (May 2019) but compare different platforms ('mobile' vs 'desktop') and or different banner types ('accessories', 'sneakers', 'clothes', 'sports_nutrition'). Just don't repeat the same test we did above 😉

Make sure you record what considerations you have for the experiment, what hypothesis test you performed ($H_0$ and $H_1$ too), and your overall conclusion/interpretation for the _business stakeholders_. Is there a follow-up you'd suggest? 

We'll be comparing the rate of conversion for users being shown the clothes banner on mobile and desktop.

Null hypothesis: There is no difference in the rate of conversion between customers on mobile and those on desktop.

Alternate hypothesis: There is a difference in the rate of conversion between customers on mobile and those on desktop.

In [56]:
df_CD = df.loc[(df["time"] >= "2019-05-01")&
               (df["product"] == "clothes")]

In [57]:
df_CD.tail()

Unnamed: 0                          order_id  \
999912     7388929  15a6cd15be985ca310f0f4fe6c186797   
999942     4124938  a6f8a9e42a4616165fae23fd7ad5bdb5   
999951     7160307  dbd36203dbf0aaee2c41920bb133294e   
999969     1305114  81a675ab9565e80aa4dcb65c1d12eb30   
999992     7735711  c40020044e86083b53c48ec23b1eda2a   

                                 user_id                           page_id  \
999912  d768f6cb066ec4cfeb92af62f1334024  20fff09ee00464f11e56082e039c5f0d   
999942  6fac446b7ead4f50c721d721f86ea28d  7a91c100705e3f21706251c980796af1   
999951  445485262fa191919fcf10829cd6e269  c47aaf1a1f20741c20d3a7a1597a776f   
999969  6c6604560d13cfd00453dc0623b39a17  063ac2781d3efff273b3f73ab7478c94   
999992  fbb6f9203a0845ab8a824855582059c3  eeae53a3fe2725c7e1d57abed2b116cb   

        product site_version                 time        title  target  
999912  clothes       mobile  2019-05-11 21:13:12  banner_show       0  
999942  clothes       mobile  2019-05-23 21:57:00  banner_show       0  
999951  clothes      desktop  2019-05-03 14:08:53  banner_show       0  
999969  clothes       mobile  2019-05-03 13:33:53  banner_show       0  
999992  clothes      desktop  2019-05-02 16:00:12  banner_show       0

In [58]:
df_CD.reset_index(inplace=True, drop=True)
df_CD.head()

Unnamed: 0                          order_id  \
0     4880034  c59e23bca9e3e74b76e5c89591920741   
1     8379527  16588be9bc5d4ef211d476e2fbadecc0   
2     4557894  7d5030b43301e95bfea4e385adb3c743   
3     8260648  51710dbb886eb764cc3b41cf4f1aa461   
4     8019572  46b3310f44d98c5b5e0296dcfee0f408   

                            user_id                           page_id  \
0  885540374a1549d83561b2a40313ff2e  503567af06c98967c3cc1b97886f9a36   
1  b77361f8d06dd11a741b5cd01ac23240  2b1fe1e4b63d6aadf4c0cf6df57b63a3   
2  f762f683aed8ab7da658d3327c0ad5f5  1f9783835a15e665aff7e554ea10345e   
3  098cee69703432819be7a116c831171c  05547166b01902fa7456641630405846   
4  6064af9e55908bd0e014d3dd20a44ee0  2c6993d9c5c7986d27ddfb6765bce9fb   

   product site_version                 time        title  target  
0  clothes      desktop  2019-05-15 13:40:11  banner_show       0  
1  clothes       mobile  2019-05-22 06:35:18  banner_show       0  
2  clothes       mobile  2019-05-08 06:37:21  banner_show       0  
3  clothes       mobile  2019-05-31 16:38:01  banner_show       0  
4  clothes       mobile  2019-05-27 06:16:13  banner_show       0

In [59]:
df_C = df_CD[df["site_version"] == "mobile"]
df_D = df_CD[df["site_version"] == "desktop"]

<ipython-input-59-09e6211482c7>:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_C = df_CD[df["site_version"] == "mobile"]
<ipython-input-59-09e6211482c7>:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_D = df_CD[df["site_version"] == "desktop"]


In [64]:
mobile_sales = df_C["target"].sum()
desktop_sales = df_D["target"].sum()
mobile_sales, desktop_sales

(1945, 758)

In [65]:
mobile_total = len(df_C)
desktop_total = len(df_D)

mobile_no_order = mobile_total - mobile_sales
desktop_no_order = desktop_total - desktop_sales

In [66]:
contigency_table2 = np.array([
    (mobile_sales, mobile_no_order),
    (desktop_sales, desktop_no_order)
])
contigency_table2

array([[ 1945, 36469],
       [  758, 14279]], dtype=int64)

In [68]:
stats.chi2_contingency(contigency_table2)

(0.007077392339641277,
 0.9329552678257845,
 1,
 array([[ 1942.58371218, 36471.41628782],
        [  760.41628782, 14276.58371218]]))

With a p-value of .93, we cannot refute the null hypothesis or prove correlation between desktop/mobile and rate of conversion.